File: security/src/test/java/top/bettercode/summer/security/SecurityErrorTest.java
package top.bettercode.summer.security;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;

/**
 * @author Peter Wu
 * @since 1.0.0
 */
@ExtendWith(SpringExtension.class)
@SpringBootTest(classes = TestApplication.class, properties = {
    "summer.web.ok-enable=false",
}, webEnvironment = RANDOM_PORT)
public class SecurityErrorTest {

  @Autowired
  TestRestTemplate restTemplate;
  final String username = "root";

  @BeforeEach
  public void setUp() {
  }

  @Test
  public void unAuth() {
    HttpHeaders httpHeaders = new HttpHeaders();
    ResponseEntity<String> entity = restTemplate
        .exchange("/testDefaultAuth", HttpMethod.GET, new HttpEntity<>(httpHeaders),
            String.class);
    System.err.println(entity.getBody());
    assertEquals(HttpStatus.UNAUTHORIZED, entity.getStatusCode());
  }


  @Test
  public void noToken() {
    HttpHeaders httpHeaders = new HttpHeaders();
//    httpHeaders.setAccept(Collections.singletonList(MediaType.APPLICATION_XML));
    ResponseEntity<String> entity = restTemplate
        .exchange("/testDefaultAuth", HttpMethod.GET, new HttpEntity<>(httpHeaders),
            String.class);
    System.err.println(entity.getBody());
    assertEquals(HttpStatus.UNAUTHORIZED, entity.getStatusCode());
  }

  @Test
  public void invalidToken() {
    HttpHeaders httpHeaders = new HttpHeaders();
//    httpHeaders.setAccept(Collections.singletonList(MediaType.APPLICATION_XML));
    ResponseEntity<String> entity = restTemplate
        .exchange("/testDefaultAuth?access_token=xxx", HttpMethod.GET, new HttpEntity<>(httpHeaders),
            String.class);
    System.err.println(entity.getBody());
    assertEquals(HttpStatus.UNAUTHORIZED, entity.getStatusCode());
  }

//  @Test
  public void expiredToken() {
    HttpHeaders httpHeaders = new HttpHeaders();
    ResponseEntity<String> entity = restTemplate
        .exchange(
            "/testDefaultAuth?access_token=",
            HttpMethod.GET, new HttpEntity<>(httpHeaders),
            String.class);
    System.err.println(entity.getBody());
    assertEquals(HttpStatus.UNAUTHORIZED, entity.getStatusCode());
  }

  @Test
  public void accessTokenError() throws Exception {
    MultiValueMap<String, Object> params = new LinkedMultiValueMap<>();
    params.add("grant_type", "password");
    params.add("scope", "trust");
    params.add("username", username);
    params.add("password", "wrong password");
    params.add("openId", "");

    ResponseEntity<String> entity = restTemplate
        .postForEntity("/oauth/token", new HttpEntity<>(params),
            String.class);
    System.err.println(entity.getBody());
    Thread.sleep(5 * 1000);
    assertEquals(HttpStatus.BAD_REQUEST, entity.getStatusCode());
  }
}

File: security/src/test/kotlin/top/bettercode/summer/security/SecurityErrorTest.kt
package top.bettercode.summer.security

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.util.LinkedMultiValueMap
import org.springframework.util.MultiValueMap

/**
 * @author Peter Wu
 * @since 1.0.0
 */
@ExtendWith(SpringExtension::class)
@SpringBootTest(classes = [TestApplication::class], properties = ["summer.web.ok-enable=false"], webEnvironment = WebEnvironment.RANDOM_PORT)
class SecurityErrorTest {
    @Autowired
    lateinit var restTemplate: TestRestTemplate
    val username = "root"

    @BeforeEach
    fun setUp() {
    }

    @Test
    fun unAuth() {
        val httpHeaders = HttpHeaders()
        val entity = restTemplate
                .exchange("/testDefaultAuth", HttpMethod.GET, HttpEntity<Any>(httpHeaders),
                        String::class.java)
        System.err.println(entity.body)
        Assertions.assertEquals(HttpStatus.UNAUTHORIZED, entity.statusCode)
    }

    @Test
    fun noToken() {
        val httpHeaders = HttpHeaders()
        //    httpHeaders.setAccept(Collections.singletonList(MediaType.APPLICATION_XML));
        val entity = restTemplate
                .exchange("/testDefaultAuth", HttpMethod.GET, HttpEntity<Any>(httpHeaders),
                        String::class.java)
        System.err.println(entity.body)
        Assertions.assertEquals(HttpStatus.UNAUTHORIZED, entity.statusCode)
    }

    @Test
    fun invalidToken() {
        val httpHeaders = HttpHeaders()
        //    httpHeaders.setAccept(Collections.singletonList(MediaType.APPLICATION_XML));
        val entity = restTemplate
                .exchange("/testDefaultAuth?access_token=xxx", HttpMethod.GET,
                        HttpEntity<Any>(httpHeaders),
                        String::class.java)
        System.err.println(entity.body)
        Assertions.assertEquals(HttpStatus.UNAUTHORIZED, entity.statusCode)
    }

//    @Test
    fun expiredToken() {
        val httpHeaders = HttpHeaders()
        val entity = restTemplate
                .exchange(
                        "/testDefaultAuth?access_token=",
                        HttpMethod.GET, HttpEntity<Any>(httpHeaders),
                        String::class.java)
        System.err.println(entity.body)
        Assertions.assertEquals(HttpStatus.UNAUTHORIZED, entity.statusCode)
    }

    @Test
    fun accessTokenError() {
        val params: MultiValueMap<String, Any> = LinkedMultiValueMap()
        params.add("grant_type", "password")
        params.add("scope", "trust")
        params.add("username", username)
        params.add("password", "wrong password")
        params.add("openId", "")
        val entity = restTemplate
                .postForEntity("/oauth/token", HttpEntity(params),
                        String::class.java)
        System.err.println(entity.body)
//        Thread.sleep((5 * 1000).toLong())
        Assertions.assertEquals(HttpStatus.BAD_REQUEST, entity.statusCode)
    }
}
